### PR TITLE
add before and after events filter

### PR DESCRIPTION
**Pull request by louis-she**
Description:

add before and after events filter

```python
@trainer.on(Events.EPOCH_COMPLETED(before=10))
def foo1():
    pass

@trainer.on(Events.EPOCH_COMPLETED(after=10))
def foo2():
    pass
```

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

**Comment by louis-she** [comment]
I removed the `WIP` in title. But one thing left is that should we add a `__and__` to the Event? Like this:

```python
@trainer.on(Events.EPOCH_COMPLETED(after=10) & Events.EPOCH_COMPLETED(before=30))
def foo1():
    # called on epoch > 10 and < 30
    pass
```

Currently with `|` or `__or__`, I think we can not achieve this with event filter only

File: tests/ignite/engine/test_custom_events.py [hunk 1]
```diff
@@ -531,4 +583,5 @@ def execute_some_handler(e):
 
     _test(Events.ITERATION_STARTED(once=1) | Events.ITERATION_STARTED(once=1), [1, 1])
     _test(Events.ITERATION_STARTED(once=1) | Events.ITERATION_STARTED(once=10), [1, 10])
-    _test(Events.ITERATION_STARTED(once=1) | Events.ITERATION_STARTED(every=3), [1, 3, 6, 9, 12, 15])
```
Comment (sadra-barikbin) [inline tests/ignite/engine/test_custom_events.py:534]: Why have you deleted the statement testing `every`?

**Review by sadra-barikbin — CHANGES_REQUESTED** [review]

**Comment by sadra-barikbin** [comment]
> I think it's ready to be reviewed. One thing to be discussed is that should we add a `__and__` operator to the Event? Like this:
> 
> ```python
> @trainer.on(Events.EPOCH_COMPLETED(after=10) & Events.EPOCH_COMPLETED(before=30))
> def foo1():
>     # called on epoch > 10 and < 30
>     pass
> ```
> 
> Currently with `|` or `__or__`, I think we can not achieve this with event filter only

We could accept both `before` and `after` in the `CallableEventWithFilter.__call__` and use `self.before_and_after_event_filter(before: Optional[int], after: Optional[int])` internally.

**Comment by louis-she** [comment]
> We could accept both before and after in the CallableEventWithFilter.__call__ and use self.before_and_after_event_filter(before: Optional[int], after: Optional[int]) internally.

hi @sadra-barikbin , you mean if a user gives `before=30 | after=10`, the internal will do the `before_and_after_event_filter`? I'm not sure about this as it could lead to some misunderstanding. Still, I think adding a `&` operator is the way to go, what do you think?

File: tests/ignite/engine/test_custom_events.py (hunk 1 above)
Comment (louis-she) [reply tests/ignite/engine/test_custom_events.py:534]: Sorry, it's an accident, fixed.

**Comment by sadra-barikbin** [comment]
> > We could accept both before and after in the CallableEventWithFilter.**call** and use self.before_and_after_event_filter(before: Optional[int], after: Optional[int]) internally.
> 
> For the `called on epoch > 10 and < 30` scenario, it'll work. Another possible scenario is `called every x epochs that after y epochs`. What do you think?

I think introducing `&` operator should be discussed in terms of its usefulness, but just in the case of mixing `before` and `after` we can do it as follows:
```python
@trainer.on(Events.EPOCH_COMPLETED(after=10, before=30))
def foo1():
    # called on epoch > 10 and < 30
    pass
```
@vfdev-5 , What's your thought?

**Comment by vfdev-5** [comment]
> 

Yes, I agree. We can just allow after and before args together and thus do not need to introduce complicated & operator.

**Comment by louis-she** [comment]
comments fixed and rebased, ready to review again @sadra-barikbin

**Comment by vfdev-5** [comment]
@louis-she could you please fix mypy issues: https://github.com/pytorch/ignite/actions/runs/3180145927/jobs/5183390359#step:12:12

**Review by vfdev-5 — APPROVED** [review]
Thanks @louis-she , looks good to me !

File: tests/ignite/engine/test_custom_events.py [hunk 2]
```diff
@@ -296,6 +313,79 @@ def test_every_event_filter_with_engine():
...
+        def _before_and_after_event():
+            nonlocal num_calls
+            num_calls += 1
+            assert getattr(engine.state, event_attr) > after
+
+        engine.run(data, max_epochs=5)
+        assert num_calls == expect_calls
+
+    _test(Events.ITERATION_STARTED, "iteration", 300, 100, 199)
+    _test(Events.EPOCH_COMPLETED, "epoch", 4, 1, 2)
```
Comment (sadra-barikbin) [inline tests/ignite/engine/test_custom_events.py:386]: Thanks @louis-she , You can improve this test by using PyTest parameterization feature as follows:
```suggestion
@pytest.mark.parametrize(
    "event_name, before, after, expected_calls",
    [
        (Events.ITERATION_STARTED, 300, 100, 199),
        (Events.EPOCH_COMPLETED, 4, 1, 2)
    ],
)
def test_before_and_after_event_filter_with_engine(event_name, before, after, expected_calls):

    data = range(100)
    engine = Engine(lambda e, b: 1)
    num_calls = 0

   @engine.on(event_name(before=before, after=after))
   def _before_and_after_event():
        nonlocal num_calls
        num_calls += 1
        assert engine.state.get_event_attrib_value(event_name) > after

    engine.run(data, max_epochs=5)
    assert num_calls == expected_calls
```

File: tests/ignite/engine/test_custom_events.py [hunk 3]
```diff
@@ -296,6 +313,79 @@ def test_every_event_filter_with_engine():
...
+    _test(Events.ITERATION_COMPLETED, "iteration", 0, 0)
+    _test(Events.ITERATION_COMPLETED, "iteration", 300, 299)
+    _test(Events.ITERATION_COMPLETED, "iteration", 501, 500)
+
+    _test(Events.EPOCH_COMPLETED, "epoch", 0, 0)
+    _test(Events.EPOCH_COMPLETED, "epoch", 3, 2)
+    _test(Events.EPOCH_COMPLETED, "epoch", 6, 5)
+
+
+def test_after_event_filter_with_engine():
```
Comment (sadra-barikbin) [inline tests/ignite/engine/test_custom_events.py:342]: An also this test.

File: tests/ignite/engine/test_custom_events.py [hunk 4]
```diff
@@ -296,6 +313,79 @@ def test_every_event_filter_with_engine():
     _test_every_event_filter_with_engine()
 
 
+def test_before_event_filter_with_engine():
```
Comment (sadra-barikbin) [inline tests/ignite/engine/test_custom_events.py:316]: And this test

**Review by sadra-barikbin — APPROVED** [review]

**Pull request merged**